plugin/core/windows.py

```diff
@@ -343,144 +90,177 @@ def __init__(
         self._settings = settings
         self._configs = configs
         self.diagnostics = diagnostics
-        self.documents = documents
         self.server_panel_factory = server_panel_factory
-        self._sessions = dict()  # type: Dict[str, List[Session]]
-        self._next_initialize_views = list()  # type: List[ViewLike]
+        self._sessions = WeakSet()  # type: WeakSet[Session]
         self._sublime = sublime
-        self._restarting = False
-        self._is_closing = False
-        self._initialization_lock = threading.Lock()
         self._workspace = workspace
+        self._pending_listeners = deque()  # type: Deque[AbstractViewListener]
+        self._listeners = WeakSet()  # type: WeakSet[AbstractViewListener]
+        self._new_listener = None  # type: Optional[AbstractViewListener]
+        self._new_session = None  # type: Optional[Session]
 
-    def on_load_project(self) -> None:
+    def on_load_project_async(self) -> None:
         # TODO: Also end sessions that were previously enabled in the .sublime-project, but now disabled or removed
         # from the .sublime-project.
+        self.end_sessions_async()
         self._configs.update()
         workspace_folders = self._workspace.update()
-        for sessions in self._sessions.values():
-            for session in sessions:
-                session.update_folders(workspace_folders)
-
-    def on_pre_close_project(self) -> None:
-        self.end_sessions()
+        for session in self._sessions:
+            session.update_folders(workspace_folders)
+
+    def register_listener(self, listener: AbstractViewListener) -> None:
+        sublime.set_timeout_async(lambda: self.register_listener_async(listener))
```

Comment:
I'll complain about method ordering again - it would be so much easier to follow the flow if this would be just after `enable_config_async`. And similarly for many other new methods.

Reply:
I moved enabled_config_async and disable_config_async up https://github.com/sublimelsp/LSP/pull/1069/commits/efecf49a5222ec8e38a79907ba93e66db3a4fdac